webhooks/elbv2/ingressclassparams_validator.go

```diff
@@ -89,6 +92,63 @@ func validateCIDR(cidr string, fieldPath *field.Path) field.ErrorList {
 	return allErrs
 }
 
+// checkSecurityGroupsSelectors will check for valid SubnetSelectors
+func (v *ingressClassParamsValidator) checkSecurityGroupsSelectors(icp *elbv2api.IngressClassParams) (allErrs field.ErrorList) {
+	if icp.Spec.SecurityGroups != nil {
+		securityGroups := icp.Spec.SecurityGroups
+		fieldPath := field.NewPath("spec", "securityGroups")
+		if securityGroups.IDs == nil && !securityGroups.ManagedInbound && securityGroups.Tags == nil {
+			allErrs = append(allErrs, field.Required(fieldPath, "must have `ids`, `managed`, or `tags`"))
+			return allErrs
+		}
+		if securityGroups.IDs != nil {
+			if len(icp.Spec.InboundCIDRs) > 0 {
+				allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "inboundCIDRs"), "May not have both `inboundCIDRs` and `securityGroups.ids`"))
+			}
+			if securityGroups.ManagedInbound {
+				allErrs = append(allErrs, field.Forbidden(fieldPath.Child("managedInbound"), "may not have both `ids` and `managedInbound` set"))
+				return allErrs
+			}
+			if securityGroups.Tags != nil {
+				allErrs = append(allErrs, field.Forbidden(fieldPath.Child("tags"), "may not have both `ids` and `tags` set"))
```

Comment:
I'm just wondering why would the `ids` and `tags` be exclusive? Let's say if a user specifies both (for fine-grained filtering purpose, or just by accident), can we just filter the SG by both field instead of error out?

Reply:
What is the use case for specifying both ids and tags? An admin would typically have one way of communicating security groups from the IaC provisioning system to the deploy-to-Kubernetes system. Either they'd propagate a list of IDs or they'd provision the groups with tags according to some convention. Why would someone use one mechanism for some groups and another mechanism for others?

Reply:
I don't see much point in having both. If you have id, there isn't much value in having the name (e.g. for the sake of documentation one can always put this information in comments or in some non-actionable annotation). Allowing both means you either have to make one take precedence when these don't agree; or still fail at runtime - but this is way more complicated to explain what's wrong (e.g. the message would have to read something like "sg-123 has name 'name-1', but the annotation specified 'name-2' - you have to either only use name/id; or if both are used - they should point to the same entity" - but having such a reasonable message is somewhat a security concern because you can now map security-group-id to a name which might have sensitive info)

At the same time clearly having a behavior to fail when both are specified allows using things like admission controllers to explicitly fail when resource are provisioned, rather than at runtime, so you can't even deploy incorrectly configured ingress.